CesiumGltf/src/Model.cpp

```diff
@@ -208,3 +211,366 @@ void Model::merge(Model&& rhs) {
+      count);
+
+  // Add the triangle's normal to each of it's vertex's accumulated normal.
+  auto addTriangleNormalToVertexNormals =
+      [&positionView,
+       &normals](TIndex tIndex0, TIndex tIndex1, TIndex tIndex2) -> void {
+    size_t index0 = static_cast<size_t>(tIndex0);
+    size_t index1 = static_cast<size_t>(tIndex1);
+    size_t index2 = static_cast<size_t>(tIndex2);
+
+    const glm::vec3& vertex0 = positionView[static_cast<int64_t>(index0)];
+    const glm::vec3& vertex1 = positionView[static_cast<int64_t>(index1)];
+    const glm::vec3& vertex2 = positionView[static_cast<int64_t>(index2)];
+
+    glm::vec3 triangleNormal = glm::cross(vertex1 - vertex0, vertex2 - vertex0);
+
+    // Add the triangle normal to each vertex's accumulated normal. At the end
+    // we will normalize the accumulated vertex normals to average.
+    normals[index0] += triangleNormal;
+    normals[index1] += triangleNormal;
+    normals[index2] += triangleNormal;
+  };
+
+  int64_t numIndices;
+  std::function<TIndex(int64_t)> getIndex;
+
+  // In the indexed case, the positions are accessed with the
+  // indices from the indexView. Otherwise, the elements are
+  // accessed directly.
+  if (indexAccessor) {
+    CesiumGltf::AccessorView<TIndex> indexView(gltf, *indexAccessor);
+    if (indexView.status() != AccessorViewStatus::Valid) {
+      return;
+    }
+    numIndices = indexView.size();
+    getIndex = [&indexView](int64_t index) { return indexView[index]; };
+  } else {
+    numIndices = int64_t(count);
+    getIndex = [](int64_t index) { return static_cast<TIndex>(index); };
+  }
+
+  switch (primitive.mode) {
+  case MeshPrimitive::Mode::TRIANGLES:
+    for (int64_t i = 2; i < numIndices; i += 3) {
+      TIndex index0 = getIndex(i - 2);
+      TIndex index1 = getIndex(i - 1);
+      TIndex index2 = getIndex(i);
+
+      addTriangleNormalToVertexNormals(index0, index1, index2);
+    }
+    break;
+
+  case MeshPrimitive::Mode::TRIANGLE_STRIP:
+    for (int64_t i = 0; i < numIndices - 2; ++i) {
+      TIndex index0;
+      TIndex index1;
+      TIndex index2;
+
+      if (i % 2) {
+        index0 = getIndex(i);
+        index1 = getIndex(i + 2);
+        index2 = getIndex(i + 1);
+      } else {
+        index0 = getIndex(i);
+        index1 = getIndex(i + 1);
+        index2 = getIndex(i + 2);
+      }
+
+      addTriangleNormalToVertexNormals(index0, index1, index2);
+    }
+    break;
+
+  case MeshPrimitive::Mode::TRIANGLE_FAN:
+    if (numIndices < 3) {
+      return;
+    }
+
+    {
+      TIndex index0 = getIndex(0);
+      for (int64_t i = 2; i < numIndices; ++i) {
+        TIndex index1 = getIndex(i - 1);
+        TIndex index2 = getIndex(i);
+
+        addTriangleNormalToVertexNormals(index0, index1, index2);
+      }
+    }
+    break;
+
+  default:
+    return;
+  }
+
+  // normalizes the accumulated vertex normals
+  for (size_t i = 0; i < count; ++i) {
+    normals[i] = glm::normalize(normals[i]);
```

Comment:
`NaN` would be the best case. Crashes _might_ be the result. It's tremendously hard to find reliable information there. E.g. https://github.com/g-truc/glm/blob/master/manual.md#section7_10 says that it ~"might crash", and **explicitly** mentions the case of `normalize` for a vector with length 0.0. **But** we didn't see crashes for that case until now (but only `NaN`). All this might well depend on some compiler settings or so...